### PR TITLE
neo wrist

### PR DESCRIPTION
**Pull request by YoniKiriaty**

### CODE REVIEW

File: src/main/java/edu/greenblitz/robotName/subsystems/arm/EndEffector/WristUtils/NeoWrist/NeoWrist.java [hunk 1]
```diff
@@ -21,21 +17,26 @@ public class NeoWrist implements IWrist {
...
 
     public NeoWrist() {
-        motor = new GBSparkMax(NeoWristConstants.PID_SLOT, CANSparkMaxLowLevel.MotorType.kBrushless);
+        motor = new GBSparkMax(NeoWristConstants.MOTOR_ID, CANSparkMaxLowLevel.MotorType.kBrushless);
         motor.config(NeoWristConstants.WRIST_CONFIG_OBJECT);
 
+        motor.getPIDController().setFeedbackDevice(motor.getEncoder());
+
         motor.getAbsoluteEncoder(SparkMaxAbsoluteEncoder.Type.kDutyCycle).setPositionConversionFactor(WristConstants.ABSOLUTE_POSITION_CONVERSION_FACTOR);
         motor.getAbsoluteEncoder(SparkMaxAbsoluteEncoder.Type.kDutyCycle).setVelocityConversionFactor(WristConstants.ABSOLUTE_VELOCITY_CONVERSION_FACTOR);
```
Comment (YoniKiriaty) [inline src/main/java/edu/greenblitz/robotName/subsystems/arm/EndEffector/WristUtils/NeoWrist/NeoWrist.java:26]: what???

File: src/main/java/edu/greenblitz/robotName/subsystems/arm/EndEffector/WristUtils/NeoWrist/NeoWristConstants.java [hunk 2]
```diff
@@ -16,12 +16,15 @@ public class NeoWristConstants {
 
     public static final PIDObject NEO_PID = new PIDObject().withKp(0.8).withKd(0.3).withKi(0).withMaxPower(1);
```
Comment (YoniKiriaty) [inline src/main/java/edu/greenblitz/robotName/subsystems/arm/EndEffector/WristUtils/NeoWrist/NeoWristConstants.java:17]: why neo pid its in calss neo

File: src/main/java/edu/greenblitz/robotName/subsystems/arm/EndEffector/WristUtils/NeoWrist/NeoWristConstants.java [hunk 3]
```diff
@@ -16,12 +16,15 @@ public class NeoWristConstants {
```
Comment (YoniKiriaty) [inline src/main/java/edu/greenblitz/robotName/subsystems/arm/EndEffector/WristUtils/NeoWrist/NeoWristConstants.java:16]: change motor id to 4

File: src/main/java/edu/greenblitz/robotName/subsystems/arm/EndEffector/WristUtils/WristConstants.java [hunk 4]
```diff
@@ -28,9 +28,7 @@ public enum PresetPositions {
 
     public static final Rotation2d FORWARD_ANGLE_LIMIT = Rotation2d.fromDegrees(200);
 
-    public static final double RELATIVE_POSITION_CONVERSION_FACTOR = 0.0328;
-
-    public static final double RELATIVE_VELOCITY_CONVERSION_FACTOR = 0.0030332432;
+    public static final double RELATIVE_CONVERSION_FACTOR = 0.0328;
 
     public static final double ABSOLUTE_POSITION_CONVERSION_FACTOR = 0.00302;
```
Comment (YoniKiriaty) [inline src/main/java/edu/greenblitz/robotName/subsystems/arm/EndEffector/WristUtils/WristConstants.java:33]: why you need that

File: src/main/java/edu/greenblitz/robotName/subsystems/arm/EndEffector/WristUtils/WristConstants.java [hunk 5]
```diff
@@ -28,9 +28,7 @@ public enum PresetPositions {
 
     public static final Rotation2d FORWARD_ANGLE_LIMIT = Rotation2d.fromDegrees(200);
 
-    public static final double RELATIVE_POSITION_CONVERSION_FACTOR = 0.0328;
-
-    public static final double RELATIVE_VELOCITY_CONVERSION_FACTOR = 0.0030332432;
+    public static final double RELATIVE_CONVERSION_FACTOR = 0.0328;
```
Comment (YoniKiriaty) [inline src/main/java/edu/greenblitz/robotName/subsystems/arm/EndEffector/WristUtils/WristConstants.java:32]: change to conversion factor if its the only one who used

File: src/main/java/edu/greenblitz/robotName/subsystems/arm/EndEffector/WristUtils/NeoWrist/NeoWrist.java [hunk 6]
```diff
@@ -21,21 +17,23 @@ public class NeoWrist implements IWrist {
...
+        motor.getPIDController().setFeedbackDevice(motor.getEncoder());
 
-        motor.getEncoder().setPositionConversionFactor(RELATIVE_POSITION_CONVERSION_FACTOR);
-        motor.getEncoder().setVelocityConversionFactor(RELATIVE_VELOCITY_CONVERSION_FACTOR);
+        motor.getEncoder().setPositionConversionFactor(CONVERSION_FACTOR);
+        motor.getEncoder().setVelocityConversionFactor(CONVERSION_FACTOR);
 
-        motor.getPIDController().setFeedbackDevice(motor.getAbsoluteEncoder(SparkMaxAbsoluteEncoder.Type.kDutyCycle));
         motor.enableSoftLimit(CANSparkMax.SoftLimitDirection.kForward, true);
         motor.enableSoftLimit(CANSparkMax.SoftLimitDirection.kReverse, true);
```
Comment (NoamRosenberg08) [inline src/main/java/edu/greenblitz/robotName/subsystems/arm/EndEffector/WristUtils/NeoWrist/NeoWrist.java:29]: change the 'true' of the is enabled of the limit to be 2 constants, one for the forward and one for the backward

File: src/main/java/edu/greenblitz/robotName/subsystems/arm/EndEffector/WristUtils/NeoWrist/NeoWrist.java [hunk 7]
```diff
@@ -63,6 +61,11 @@ public void moveToAngle(Rotation2d targetAngle) {
         motor.getPIDController().setReference(targetAngle.getRadians(), CANSparkMax.ControlType.kPosition, PID_SLOT, motor.getPIDController().getFF());
```
Comment (NoamRosenberg08) [inline src/main/java/edu/greenblitz/robotName/subsystems/arm/EndEffector/WristUtils/NeoWrist/NeoWrist.java:61]: use simple motor feed forward for FF. the getFF of the motorws PIDController is for a constant FF (like only kS)

**Review by NoamRosenberg08 — CHANGES_REQUESTED** [review]

**Review by yonnbro — APPROVED** [review]
aside from others comments - approved

**Review by bossyshrimp68 — APPROVED** [review]
what noam said, but besides that gg

File: src/main/java/edu/greenblitz/robotName/subsystems/arm/wrist/SimulationWrist/SimulationWrist.java [hunk 8]
```diff
@@ -68,6 +68,11 @@ public void moveToAngle(Rotation2d targetAngle) {
         setVoltage(controller.calculate(wristSimulation.getAngleRads()));
     }
 
+    @Override
+    public void standInPlace() {
+        setVoltage(SimulationWristConstants.SIMULATION_FEED_FORWARD.calculate(0));
```
Comment (NoamRosenberg08) [inline src/main/java/edu/greenblitz/robotName/subsystems/arm/wrist/SimulationWrist/SimulationWrist.java:73]: again, multiply by the cosine of the angle

File: src/main/java/edu/greenblitz/robotName/subsystems/arm/wrist/WristConstants.java [hunk 9]
```diff
@@ -28,13 +28,7 @@ public enum PresetPositions {
...
-    public static final double RELATIVE_POSITION_CONVERSION_FACTOR = 0.0328;
-
-    public static final double RELATIVE_VELOCITY_CONVERSION_FACTOR = 0.0030332432;
-
-    public static final double ABSOLUTE_POSITION_CONVERSION_FACTOR = 0.00302;
-
-    public static final double ABSOLUTE_VELOCITY_CONVERSION_FACTOR = 0.00302;
+    public static final double CONVERSION_FACTOR = 0.0328;
 
     public static final double kS = 1;
```
Comment (NoamRosenberg08) [inline src/main/java/edu/greenblitz/robotName/subsystems/arm/wrist/WristConstants.java:33]: why is it here? the kS is different for each motor and sub-mechanism thus, it needs to be in each constants file of each subsystem

**Review by NoamRosenberg08 — CHANGES_REQUESTED** [review]

File: src/main/java/edu/greenblitz/robotName/subsystems/arm/wrist/NeoWrist/NeoWrist.java [hunk 10]
```diff
@@ -14,23 +15,26 @@ public class NeoWrist implements IWrist {
...
-        motor.enableSoftLimit(CANSparkMax.SoftLimitDirection.kReverse, true);
+        motor.enableSoftLimit(CANSparkMax.SoftLimitDirection.kForward, NeoWristConstants.IS_SOFT_FORWARD_LIMIT_ENABLED);
+        motor.enableSoftLimit(CANSparkMax.SoftLimitDirection.kReverse, NeoWristConstants.IS_SOFT_BACKWARD_LIMIT_ENABLED);
         motor.setSoftLimit(CANSparkMax.SoftLimitDirection.kReverse, WristConstants.BACKWARD_ANGLE_LIMIT.getRadians());
         motor.setSoftLimit(CANSparkMax.SoftLimitDirection.kForward, WristConstants.FORWARD_ANGLE_LIMIT.getRadians());
 
+        motor.getReverseLimitSwitch(SparkMaxLimitSwitch.Type.kNormallyOpen).enableLimitSwitch(NeoWristConstants.IS_BACKWARD_LIMIT_SWITCH_ENABLED);
+        motor.getForwardLimitSwitch(SparkMaxLimitSwitch.Type.kNormallyOpen).enableLimitSwitch(NeoWristConstants.IS_FORWARD_LIMIT_SWITCH_ENABLED);
+
+        resetAngle(Rotation2d.fromRotations(motor.getAbsoluteEncoder(SparkMaxAbsoluteEncoder.Type.kDutyCycle).getPosition()));
```
Comment (NoamRosenberg08) [inline src/main/java/edu/greenblitz/robotName/subsystems/arm/wrist/NeoWrist/NeoWrist.java:37]: make a function for resetEncoder

**Review by NoamRosenberg08 — CHANGES_REQUESTED** [review]

File: src/main/java/edu/greenblitz/robotName/subsystems/arm/wrist/NeoWrist/NeoWristConstants.java [hunk 11]
```diff
@@ -9,18 +9,28 @@
...
-    public static final int MOTOR_ID = 1;
+    public static final int MOTOR_ID = 4;
 
     public static final int PID_SLOT = 0;
 
-    public static final PIDObject NEO_PID = new PIDObject().withKp(0.8).withKd(0.3).withKi(0).withMaxPower(1);
+    public static final PIDObject PID = new PIDObject().withKp(0.8).withKd(0.3).withKi(0).withMaxPower(1);
 
-    public static final SimpleMotorFeedforward WRIST_FEED_FORWARD = new SimpleMotorFeedforward(WristConstants.kS, WristConstants.kV, WristConstants.kA);
+    public static final SimpleMotorFeedforward WRIST_FEED_FORWARD = new SimpleMotorFeedforward(1, 1, 1);
```
Comment (yonnbro) [inline src/main/java/edu/greenblitz/robotName/subsystems/arm/wrist/NeoWrist/NeoWristConstants.java:18]: Why don't we tuse the `ArmFeedforward` wpilib object?

File: src/main/java/edu/greenblitz/robotName/subsystems/arm/wrist/NeoWrist/NeoWrist.java [hunk 12]
```diff
@@ -53,9 +56,24 @@ public void resetAngle(Rotation2d position) {
...
+        motor.getPIDController().setReference(
+                targetAngle.getRadians(),
+                CANSparkMax.ControlType.kPosition,
+                NeoWristConstants.PID_SLOT,
+                NeoWristConstants.WRIST_FEED_FORWARD.calculate(lastInputs.velocity)
+        );
+    }
+
+    @Override
+    public void standInPlace() {
```
Comment (yonnbro) [inline src/main/java/edu/greenblitz/robotName/subsystems/arm/wrist/NeoWrist/NeoWrist.java:75]: I think the `standInPlace ` function should also do PID control on the desired position, no?

File: src/main/java/edu/greenblitz/robotName/subsystems/arm/wrist/SimulationWrist/SimulationWristConstants.java [hunk 13]
```diff
@@ -2,13 +2,16 @@
...
 public class SimulationWristConstants {
 
     public static final int NUMBER_OF_MOTORS = 1;
 
-    public static final double GEAR_RATIO = 1 / WristConstants.RELATIVE_POSITION_CONVERSION_FACTOR;
+    public static final double GEAR_RATIO = 1 / WristConstants.CONVERSION_FACTOR;
 
     public static final PIDObject SIMULATION_PID = new PIDObject().withKp(0.5).withKd(0).withMaxPower(1);
 
+    public static final SimpleMotorFeedforward SIMULATION_FEED_FORWARD = new SimpleMotorFeedforward(0,0);
```
Comment (yonnbro) [inline src/main/java/edu/greenblitz/robotName/subsystems/arm/wrist/SimulationWrist/SimulationWristConstants.java:15]: agian - why not ArmFeedForward

File: src/main/java/edu/greenblitz/robotName/subsystems/arm/wrist/NeoWrist/NeoWrist.java (hunk 12 above)
Comment (YoniKiriaty) [reply src/main/java/edu/greenblitz/robotName/subsystems/arm/wrist/NeoWrist/NeoWrist.java:75]: no.

Comment (yonnbro) [reply src/main/java/edu/greenblitz/robotName/subsystems/arm/wrist/NeoWrist/NeoWrist.java:75]: but the mechanism will not stay in place... it is subjected to external forces acting on it that will effect it

Comment (YoniKiriaty) [reply src/main/java/edu/greenblitz/robotName/subsystems/arm/wrist/NeoWrist/NeoWrist.java:75]: this is outside of the scope of this CR, making standInPlace use PID changes a lot of the code.

Comment (YoniKiriaty) [reply src/main/java/edu/greenblitz/robotName/subsystems/arm/wrist/NeoWrist/NeoWrist.java:75]: also at that point you are just doing moveToAngle

Comment (yonnbro) [reply src/main/java/edu/greenblitz/robotName/subsystems/arm/wrist/NeoWrist/NeoWrist.java:75]: please bring it up in the weekly tomorrow then

**Pull request merged**